0000-macro-value.md

```diff
@@ -0,0 +1,300 @@
+That is, `m!(array[0])` would call `Deref::deref` and return that reference directly.
+If `m!` were defined as `=> ( &mut $x );`,
+then `Deref::deref_mut` would be called.
+If the capture is expanded only once,
+this behaves identically to an `$:expr` capture,
+except for the evaluation timing of side effects.
+
+However, `$:place` adds a new concept to Rust,
+that of capturing a _place_ directly.
+This is entirely impossible in surface Rust today.
+This form of capturing may be more intuitive to macro authors
+(who are already used to and use similar behavior from `$:expr`).
+However,
+as this is much more complicated on the implementation side than a simple `$:value`,
+and `$:value` offers most of the benefit of `$:place` without the extra implementation complexity,
+`$:place` is just offered here as an alternative.
+
+## `macro fn`
+[macro_fn]: #macro-fn
+
+Another possibility that's been discussed is `macro fn`.
+Basically, these would be `fn`, and have the semantics of `fn`,
+but be duck typed (like macros) and semantically copy/pasted into the calling scope.
+This is basically the exact feature that "`macro_rules!` with function-like captures" is trying to serve,
+except for one important thing: a `macro fn` is (potentially) still a `fn` in that it has one fixed airity,
+and can't be overloaded like a macro can.
+Basically, `macro fn` is asking for "macros 2.0,"
+which is still desirable, but still a _long_ ways off.
+`$:value` offers a small improvement in the status quo without adding a completely new system into the language.
+
+# Rationale
+[rationale]: #rationale
+
+`$:value` simplifies the authoring of `macro_rules` macros,
+as authors now no longer need to learn and remember to use the `match` trick to bind macro value arguments,
+and instead can just use the `$:value` matcher to get function-argument semantics.
+Thus, while adding to the semantics provided by the Rust compiler,
+it reduces the needed complexity to write correct `macro_rules!`.
+
+Additionally,
+it is impossible to have a repetition of expr captures that has function-argument like drop timing through use of the `match` trick alone,
+as it requires knowing the airity of the captures ahead of time to name each capture.
+`$:value` directly unlocks properly and fully variadic macros that act like function calls with respect to temporary lifetimes.
+
+# Prior art
+[prior-art]: #prior-art
+
+Kotlin's [`inline fn`](https://kotlinlang.org/docs/inline-functions.html) behave similarly to (typechecked) macros.
+Semantically, `inline fn` is inlined into the call site,
+allowing things like `return`/`break`/`continue`ing from the calling scope.
+Otherwise, `inline fn` behaves semantically like a function call.
+
+[POV-Ray's macros](http://www.povray.org/documentation/3.7.0/r3_3.html#r3_3_2_8_3)
+are also interesting for comparison.
+
+The author knows of no other languages with macro-like functionality
+that isn't just textual or lexical replacement.
```

Comment:
There have been similar proposals to other languages, coming either from the point of view of macros or from the one of functions.

For instance, there was at least one C++ proposal to add the opposite to templates. That is, so that templates could also capture expressions rather than just values.